src/main/java/net/starlark/java/eval/StringModule.java

```diff
@@ -80,6 +84,28 @@ static String slice(String s, int start, int stop, int step) {
     }
   }
 
+  // Practically the most strings in Starlark are ASCII strings.
+  // This is a cache of single-char strings to avoid allocation in s[i] operation.
+  // It is a nested class to delay the initialization of the cache until it is really needed.
+  private static class AsciiCharStringsHolder {
+    private static final String[] ASCII_CHAR_STRINGS = new String[0x80];
+    static {
+      for (int i = 0; i != ASCII_CHAR_STRINGS.length; ++i) {
+        ASCII_CHAR_STRINGS[i] = String.valueOf((char) i);
+      }
+    }
+  }
+
+  // Returns s[index]
+  static String index(String s, int index) {
```

Comment:
Hoist the charAt call into the caller---in other words, make this function a memoized version of Character.toString for ASCII. Nothing is gained by doing the charAt call here, because substring always allocates a new array.  Let's call it memoizedCharAt.

Also, call it from StringModule.elems too.

Reply:
Substring allocates only since Java 9, before that it reuses the allocated array.

I know Java 8 is EOL already, but we still have not migrated from Java 8 everywhere. So I'd prefer keeping this function signature (and reuse in `elems` and maybe rename).

Reply:
Substring sharing makes this change even more important: you don't want s[i] to extend the lifetime of a big string s. (For a substring s[i:j], I think it's reasonable to expect users to understand the implications of sharing on lifetimes.)

Reply:
OK, not a big deal anyway, since all strings are practically ASCII. I'll restructure it.

> Let's call it memoizedCharAt

You probably meant `memoizedCharToString` or something.

Reply:
Agreed. Yes, your name is better.